### ISSUE TITLE
Unverified breakpoint

### ISSUE THREAD

**Issue by pjebs** [Microsoft/vscode-go#1421]
When I place a breakpoint and then run delve through the IDE,
the dots turn from red to "unverified".
The debugger doesn't realise that the breakpoints are even there.
When I exit debugger they go red again.

Proviso:
Occasionally it works as intended when I restart computer, but once it stops working, it never starts working again until a restart (which increases the chance of it working)

**Comment by pjebs** [comment]
It works perfectly when using delve independently.

**Comment by ramya-rao-a** [comment]
Can you try debugging a simple single file program and see if you get the same issue?
If not, then can you share a simple sample code where you get the issue so that I can try to get a repro?

Also try adding `"trace": "verbose"` to your debug config in the launch.json file. This will dump a series of logs in the Debug console. Share that.

**Comment by pjebs** [comment]
I think (I haven't tested) the issue is when I open my project in VSCode, I need open the bottom panel, then go to `terminal` and then change the `pwd` to the actual project directory.

VSCode doesn't automatically do that when I open a project, so `delve` get's confused and doesn't work properly (even though VSCode doesn't realise that)

**Comment by ramya-rao-a** [comment]
The terminal has nothing to do with how the Go extension runs dlv

**Comment by pjebs** [comment]
I think delve *requires* the working directory _to be_ the directory of the project. VSCode doesn't do  that automatically.

**Comment by ramya-rao-a** [comment]
VS Code sets the working directory to be what was set in the `program` attribute of the debug configuration provided in the launch.json file.

**Comment by ramya-rao-a** [comment]
@pjebs Any updates here? Are you still seeing the issue?

**Comment by pjebs** [comment]
I haven't used VSCode for a while. I'll use it again in a week or two.

**Comment by Wilkolicious** [comment]
Hi, I am getting this issue with this debug output:
```
Verbose logs are written to:
/var/folders/9n/_65b__pd6pd4cn85ttzwpy6c0000gn/T/vscode-go-debug.txt
11:23:40, 2018-1-19
InitializeRequest
InitializeResponse
Using GOPATH: /Users/Wilkolicious/go
Running: /usr/local/bin/dlv debug --headless=true --listen=127.0.0.1:2345 --log=true --
2018/01/19 11:23:51 server.go:73: Using API v1
2018/01/19 11:23:51 debugger.go:97: launching process with args: [/Users/Wilkolicious/git/microservice-provisions/src/debug]
API server listening at: 127.0.0.1:2345
InitializeEvent
SetBreakPointsRequest
All cleared
Creating on: /Users/Wilkolicious/git/microservice-provisions/src/controllers/submitkey.get.go:17
Creating on: /Users/Wilkolicious/git/microservice-provisions/src/controllers/submitkey.get.go:21
Creating on: /Users/Wilkolicious/git/microservice-provisions/src/controllers/submitkey.get.go:23
Error on CreateBreakpoint: unknown file: /Users/Wilkolicious/git/microservice-provisions/src/controllers/submitkey.get.go
Error on CreateBreakpoint: unknown file: /Users/Wilkolicious/git/microservice-provisions/src/controllers/submitkey.get.go
Error on CreateBreakpoint: unknown file: /Users/Wilkolicious/git/microservice-provisions/src/controllers/submitkey.get.go
All set:[null,null,null]
SetBreakPointsResponse
ConfigurationDoneRequest
ContinueRequest
ContinueResponse
ThreadsRequest
2018/01/19 11:23:52 debugger.go:499: continuing
Connecting to DB server...Connected
Starting Microserver...
```

I've checked the files do exist from those absolute paths, e.g.
`vim /Users/Wilkolicious/git/microservice-provisions/src/controllers/submitkey.get.go`

Is this a permissions problem?  The `Code` and `Code Helper` processes are running as me.

**Comment by pjebs** [comment]
Can you try with gdlv first. If that works, then try again with VSCode.

https://github.com/aarzilli/gdlv

**Comment by ramya-rao-a** [comment]
@pjebs Any reason you are suggesting to try gdlv instead of dlv? The Go extension uses the latter not the former.

@Wilkolicious As per your logs, the GOPATH is `/Users/Wilkolicious/go` and you are trying to set breakpoints in a file that is not under your GOPATH. Why so?

**Comment by Wilkolicious** [comment]
@ramya-rao-a Thanks for taking a look.

All my git repo's live under `/Users/Wilkolicious/git`.
My GOPATH as you've rightly identified is positioned at `/Users/Wilkolicious/go`, and thus any Go libs etc.

I have a symlink like so:
`/Users/Wilkolicious/go/src/github.com/CompanyName/microservice-provisions` -> `/Users/Wilkolicious/git/microservice-provisions`.

That is, I keep all our company's repo's in one place, and the Go files in another.

**Comment by ramya-rao-a** [comment]
@Wilkolicious We have seen the breakpoint creation failing with `unknown file:..` in case of symlink files before. I dont have answers there. https://github.com/Microsoft/vscode-go/issues/1307#issuecomment-352190192 was what one of our other users concluded.

Why not use multiple GOPATHs instead of symlinks? You can set GOPATH to `/Users/Wilkolicious/go:/Users/Wilkolicious/someotherlocation`

**Comment by pjebs** [comment]
@ramya-rao-a  `gdlv` is a gui front end for `dlv`. It attempts to do the same thing as vscode.

**Comment by jbowen93** [comment]
I'm also seeing this issue with this output:
```
Verbose logs are written to:
/var/folders/np/3_q35m8d619dy31y91vn7r0h00d86p/T/vscode-go-debug.txt
5:06:30 PM, 1/24/2018
InitializeRequest
InitializeResponse
Using GOPATH: /Users/bowenjosh/OSS/golang
Running: /Users/bowenjosh/homebrew/bin/dlv debug --headless=true --listen=127.0.0.1:2345 --log=true --
2018/01/24 17:06:49 server.go:73: Using API v1
2018/01/24 17:06:49 debugger.go:97: launching process with args: [/Users/bowenjosh/OSS/golang/src/D1/D2/D3/debug]
API server listening at: 127.0.0.1:2345
InitializeEvent
SetBreakPointsRequest
All cleared
Creating on: /Users/bowenjosh/Oss/golang/src/D1/D2/D3/pkg/server/server.go:26
Error on CreateBreakpoint: unknown file: /Users/bowenjosh/Oss/golang/src/D1/D2/D3/pkg/server/server.go
All set:[null]
SetBreakPointsResponse
ConfigurationDoneRequest
ContinueRequest
ContinueResponse
ThreadsRequest
```

Delve Info
```
Delve Debugger
Version: 1.0.0-rc.2
Build: v1.0.0-rc.2
```
VSCode Info
```
1.19.2
490ef761b76b3f3b3832eff7a588aac891e5fe80
x64
```
Go Info
```
go version go1.9.3 darwin/amd64
```

**Comment by ramya-rao-a** [comment]
@jbowen93 I am guessing you are using symlinks as well? Please see https://github.com/Microsoft/vscode-go/issues/1421#issuecomment-359127404

**Comment by DazedNConfused-** [comment]
I am having an almost identical problem: all my repos live in /home/user/git whereas the GOPATH lives in /home/user/go. 

The workaround I have applied in order to keep things tidy is what everybody else struggling with this does: symlink the project from git/ into /go.

And I am also having these 'unverified breakpoint' problems. The only ones that trigger are those in the main() entrypoint; every other just fails. What are the possible suggested solutions (that allow me to keep this separation of folders)?

**Comment by jbowen93** [comment]
@ramya-rao-a I think you're correct with the symlinks, though https://github.com/Microsoft/vscode-go/issues/1307#issuecomment-352190192 shows the issue more clearly. For some reason my `/tmp` is a symlink which appears to break things.

**Comment by DazedNConfused-** [comment]
> Sorry for repeatedly editing this comment, trying to avoid leaving red herrings.
> 
> I've now got breakpoints working if I do:
> code /Users/bowenjosh/Oss/golang/src/D1/D2/D3
> 
> They don't work if I do and of the following:
> code . with cwd being /Users/bowenjosh/Oss/golang/src/D1/D2/D3
> code ~/Oss/golang/src/D1/D2/D3
> code $HOME/Oss/golang/src/D1/D2/D3
> 
> This is good enough for me so I'm not planning oninvestigating this issue anymore but I'm happy to provide more information if it helps someone else.

This doesn't work for me. In fact, my experience is exactly like #1453, where any non-main breakpoints just turn "unverified".

I don't know if everything is related somehow. Should I create another separate issue?

**Comment by pjebs** [comment]
I tried getting dlv to work via vscode. same issue as what I reported above.

**Comment by DazedNConfused-** [comment]
I can verify that GoLand doesn't have any of these issues while attaching the debugger, so it seems to be a problem directly tied to the vscode-go plugin.

**Comment by pjebs** [comment]
gdlv works everytime too.

**Comment by ramya-rao-a** [comment]
@jbowen93 Can you provide details on what your GOPATH is and which folder have you symlinked to which location? i.e Are the physical files at `/Users/bowenjosh/Oss/golang/src/D1/D2/D3` or are they somewhere else and you have symlinked them to /Users/bowenjosh/Oss/golang/src/D1/D2/D3? At what level have done the symlink?

@DazedNConfused-

> all my repos live in /home/user/git whereas the GOPATH lives in /home/user/go. ... What are the possible suggested solutions (that allow me to keep this separation of folders)?

Why not set GOPATH to `/home/user/go:/home/user/git`?

**Comment by jbowen93** [comment]
@ramya-rao-a My GOPATH is `/Users/bowenjosh/OSS/golang` and there are no symlinks in the path to the code.

**Comment by DazedNConfused-** [comment]
@ramya-rao-a 

> Why not set GOPATH to /home/user/go:/home/user/git?

I have done exactly that by specifying the merged GOPATH in the configuration's `env` attribute inside launch.json:

```json
"env": {
                "GOPATH": "/home/user/go:/home/user/git"
            },
```

Still same results.

**Comment by ramya-rao-a** [comment]
@DazedNConfused- And you have removed the symlinks?

**Comment by DazedNConfused-** [comment]
@ramya-rao-a No, I did not. Upon removing the symlinks (replacing them with the tweaked GOPATH), non-main breakpoints start working as intended.

So it seems that the issue is tied to symlinks after all.

**Comment by ramya-rao-a** [comment]
@DazedNConfused-  For the `exec` scenario, can you try running delve directly and see if you are able to hit the non-main breakpoints?

**Comment by DazedNConfused-** [comment]
@ramya-rao-a I have tried running delve exec. Non-main breakpoints do trigger.

**Comment by ramya-rao-a** [comment]
@DazedNConfused- Please log a separate issue for the not being able to hit non main breakpoints when using `exec`. Include a sample code that I can use to repro along with logs after you set `"trace": "verbose"` in your debug config. Share the debug config too

**Comment by pjebs** [comment]
I noticed the title was changed to "when using symlinks". I am not using symlinks.

**Comment by ramya-rao-a** [comment]
@pjebs Ah, sorry about that. Symlinks turned out to be the issue for @DazedNConfused- 
Do you mind pinging me on slack or gitter so that we can try and see why things are not working for you? I am running out of ideas.

**Comment by pjebs** [comment]
We can organize to do a screenshare over skype sometime. I'm quite busy for this week. Let me know in a week time.

**Comment by sarrotnam** [comment]
Is there any solution for this yet?

**Comment by nicolinux** [comment]
I have the same problem but with a Python project. It works sometimes, and sometimes it doesn't :(

**Comment by DaveyEdwards** [comment]
I have this issue with Windows 10 debugging Javascript as well, the last week or so none of my configs that previously worked for debugging are working.  

Something to note as well is even when they were "working" it was only working about 60% of the time, to fix I would have to keep restarting the debugger

**Comment by pjebs** [comment]
@DaveyEdwards debugging javascript is not related to debugging Go using delve.

**Comment by tapir** [comment]
Any news on this? Ona freshly installed linux+vscode+go computer I'm unable to break successfully yet.

**Comment by akulich** [comment]
Having the same issue on Windows
Code v1.23.0
go version go1.10.2 windows/amd64
Delve Debugger
Version: 1.0.0
Build: $Id: c98a142125d0b17bb11ec0513bde346229b5f533 $

Project is in GOPATH

Breakpoints added before debugger starts work fine, but if I add new one during debugging or try to re-add on the same line getting 'Unverified breakpoint'

Trace says:
SetBreakPointsRequest
Clearing: 1

**Comment by buzzdan** [comment]
Any news on this issue ?

**Comment by kcramer97** [comment]
I was having the same issue and solved it by changing my launch.json file in vscode.    For my debugging I was running debugging in a docker container remotely and did the following,
1.  ADD . /dev_dir      // Copied the source code to a development directory in the container 
2. Modify you launch.json file,  The key value is "remotePath"  It is critical this is the directory that your source was copied to in the docker container 
{
	"name": "Remote",
	"type": "go",
	"request": "launch",
	"mode": "remote",
	"remotePath": "${workspaceRoot}",
	"port": 2345,
	"host": "127.0.0.1",
	"program": "${workspaceRoot}",
	"env": {}
}

As soon as I fixed remotePath all the breakpoints worked again!

**Comment by pjebs** [comment]
@kcramer97 Thank you for the suggestion. However I still think it's unnecessary to have to have Docker involved when gdlv and Goland both do it successfully. No idea why VSCode can't do it.

**Comment by pjebs** [comment]
I don't think the maintainer is actually interested in solving it. She offered to solve my issue but never carried through with it.

Jetbrains has a special offer for international friendship day: https://www.jetbrains.com/promo/friends/
If you want to get goland IDE (which definitely perfectly runs delve), you can get it for $44.50 USD. After the promition, it will become $89 USD

**Comment by ramya-rao-a** [comment]
@pjebs And here I was waiting for a ping from you based on https://github.com/Microsoft/vscode-go/issues/1421#issuecomment-365463683

All,
I cannot solve an issue I cannot repro. In such cases, the best I can do is debug the issue along with you either via slack or gitter. Everyone here is free to ping me on either of those channels, and I can help you to [set up a local environment of the Go extension](https://github.com/Microsoft/vscode-go/wiki/Building,-Debugging-and-Sideloading-the-extension-in-Visual-Studio-Code#building-and-debugging-the-extension) (which is very straight forward) and [debug the debugger](https://github.com/Microsoft/vscode-go/tree/master/src/debugAdapter) 

I most cases, the problem is with the configuration or gopath which we can get to easily by debugging the extension.

@kcramer97 Yes, you are right. `remotePath` is the key. @pjebs I believe @kcramer97's scenario was that of  remote debugging and so remotePath and containers are part of the process

@akulich One limitation of dlv is that you can only add breakpoints when the program is halted. Which means you can only add a breakpoint when another breakpoint is hit or before starting the debug session. #978 is tracking the possibility of getting the pause command to work so as that we can add breakpoints

Others referring to issues with debugging python and javascript please log an issue in the python extension and the main vscode repo respectively

**Comment by afboteros** [comment]
In my case this issue was happening because i did not add %GOPATH%\bin to PATH on Windows 10, i added it and reinstall delve with 

> go get -u github.com/derekparker/delve/cmd/dlv

Then it works.

**Comment by ramya-rao-a** [comment]
Update:

@akulich The limitation of breakpoints can only be added when the program is halted is no longer true. A few updates ago, we have fixed this limitation.

As mentioned in https://github.com/Microsoft/vscode-go/issues/1421#issuecomment-410107474, unfortuantely, there is nothing much I can do here and therefore I am closing this issue.

If anybody sees the problem of unverified breakpoints in the latest update to the extension, please log an issue with sample code and debug logs.

Happy Coding!
